packages/core/src/command/hl7v2-subscriptions/hl7v2-to-fhir-converter.ts

```diff
@@ -0,0 +1,324 @@
+            ...sharedUploadParams.errorConfig,
+            errorMessage: "Error uploading updated HL7 Combined FHIR bundle to S3",
+          },
+        }
+      : undefined),
+  });
+
+  const bundlePresignedUrlPromise = shouldNotifyApi
+    ? getPresignedUrl({
+        s3Client,
+        bucketName,
+        fileName: newBundleFileName,
+        cxId,
+        messageId,
+        timestamp: messageReceivedTimestamp,
+        messageType: msgType,
+        patientId,
+        log,
+      })
+    : Promise.resolve("");
+
+  const [, , bundlePresignedUrl] = await Promise.all([
+    newBundleUploadPromise,
+    combinedBundleUploadPromise,
+    bundlePresignedUrlPromise,
+  ]);
+
+  if (shouldNotifyApi && bundlePresignedUrl) {
+    try {
+      await executeWithNetworkRetries(
+        async () =>
+          await axios.post(internalHl7RouteUrl, undefined, {
+            params: {
+              cxId,
+              patientId,
+              presignedUrl: bundlePresignedUrl,
+            },
+          })
+      );
+      log(`Successfully sent HL7 FHIR bundle to the API`);
+    } catch (err) {
+      log(`Error hitting the API endpoint: - ${errorToString(err)}`);
+      throw err;
+    }
+  } else {
+    log(`Skipping API notification for message type: ${msgType.triggerEvent}`);
+  }
+}
+
+async function saturateConvertedBundle({
+  bundle,
+  patientId,
+  cxId,
+  messageId,
+  messageReceivedTimestamp,
+  internalGetPatientUrl,
+  log,
+}: {
+  bundle: Bundle<Resource>;
+  patientId: string;
+  cxId: string;
+  messageId: string;
+  messageReceivedTimestamp: string;
+  internalGetPatientUrl: string;
+  log: typeof console.log;
+}) {
+  try {
+    const patient = await executeWithNetworkRetries(
+      async () => await axios.get(internalGetPatientUrl)
+    );
+    const fhirPatient = toFhirPatient({ id: patientId, data: patient.data });
+    const fhirPatientEntry = buildBundleEntry(fhirPatient);
+    const combinedEntries = bundle.entry ? [fhirPatientEntry, ...bundle.entry] : [];
+    return buildBundle({ type: "collection", entries: combinedEntries });
+  } catch (error) {
+    const msg = "Error retrieving patient data on HL7-to-FHIR conversion";
+    log(`${msg}: ${errorToString(error)}`);
+    capture.message(msg, {
+      extra: {
+        patientId,
+        cxId,
+        messageId,
+        timestamp: messageReceivedTimestamp,
+      },
+    });
+
+    throw error;
+  }
+}
+
+async function mergeIncomingBundleIntoCombined({
+  cxId,
+  patientId,
+  existingBundle,
+  incomingBundle,
+  log,
+}: {
+  cxId: string;
+  patientId: string;
+  existingBundle: Bundle<Resource>;
+  incomingBundle: Bundle<Resource>;
+  log: typeof console.log;
+}) {
+  try {
+    const combinedEntries = [...(existingBundle.entry || []), ...(incomingBundle.entry || [])];
+    const combinedBundle = buildBundle({
+      type: "collection",
+      entries: combinedEntries,
+    });
+
+    const dedupedCombinedBundle = await deduplicate({
```

Comment:
i remember hearing that our dedupe logic was pretty inefficient or slow right now. That's purely for computational reasons, right? These bundles will be small, but any idea what the performance will look like? ballpark est of 3, 30 or 300ms? If the lambdas are going to take 300ms+ to do the deduplicate that's good for me to keep in mind